qt/Torrent.cc

```diff
@@ -173,228 +108,262 @@ Torrent::KeyList const Torrent::detailStatKeys{
-        changed = true;
+        changed = !setme.isEmpty();
+        setme.clear();
+        return changed;
     }
 
-    return changed;
+    return change(setme, QString::fromUtf8(str, len));
 }
 
-bool Torrent::setString(int i, char const* value, size_t len)
+bool change(Peer& setme, tr_variant* value)
 {
     bool changed = false;
 
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::String);
-
-    auto& oldval = myValues[i];
-    auto const newval = QString::fromUtf8(value, len);
-
-    if (oldval != newval)
+    size_t pos = 0;
+    tr_quark key;
+    tr_variant* child;
+    while (tr_variantDictChild(value, pos++, &key, &child))
     {
-        oldval = newval;
-        changed = true;
+        switch (key)
+        {
+#define HANDLE_KEY(key) case TR_KEY_ ## key: \
+    changed |= change(setme.key, child); break;
+
+            HANDLE_KEY(address)
+            HANDLE_KEY(clientIsChoked)
+            HANDLE_KEY(clientIsInterested)
+            HANDLE_KEY(clientName)
+            HANDLE_KEY(flagStr)
+            HANDLE_KEY(isDownloadingFrom)
+            HANDLE_KEY(isEncrypted)
+            HANDLE_KEY(isIncoming)
+            HANDLE_KEY(isUploadingTo)
+            HANDLE_KEY(peerIsChoked)
+            HANDLE_KEY(peerIsInterested)
+            HANDLE_KEY(port)
+            HANDLE_KEY(progress)
+            HANDLE_KEY(rateToClient)
+            HANDLE_KEY(rateToPeer)
+#undef HANDLE_KEY
+        default:
+            break;
+        }
     }
 
     return changed;
 }
 
-bool Torrent::setIcon(int i, QIcon const& value)
-{
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::Icon);
-
-    myValues[i].setValue(value);
-    return true;
-}
-
-int Torrent::getInt(int i) const
+bool change(TorrentFile& setme, tr_variant* value)
 {
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::Int);
-    // assert(!myValues[i].isNull());
-
-    return myValues[i].toInt();
-}
+    bool changed = false;
 
-time_t Torrent::getTime(int i) const
-{
-    assert(0 <= i && i < PROPERTY_COUNT);
-    assert(myProperties[i].type == QVariant::DateTime);
-    // assert((i == DATE_ADDED) || !myValues[i].isNull());
+    size_t pos = 0;
+    tr_quark key;
+    tr_variant* child;
+    while (tr_variantDictChild(value, pos++, &key, &child))
+    {
+        switch (key)
+        {
+#define HANDLE_KEY(key) case TR_KEY_ ## key: \
+    changed |= change(setme.key, child); break;
+
+            HANDLE_KEY(have)
+            HANDLE_KEY(priority)
+            HANDLE_KEY(wanted)
+#undef HANDLE_KEY
+#define HANDLE_KEY(key, field) case TR_KEY_ ## key: \
+    changed |= change(setme.field, child); break;
+
+            HANDLE_KEY(bytesCompleted, have)
+            HANDLE_KEY(length, size)
+            HANDLE_KEY(name, filename)
+#undef HANDLE_KEY
+        default:
+            std::cerr << "unhandled " << tr_quark_get_string(key, nullptr) << std::endl;
```

Comment:
Why print here and not in other cases (like `Peer` above or `TrackerStat` below)? Should be consistent I guess, one way or another.

Reply:
Good catch! TBH it's a leftover -- I had these in the rest of the code as well while testing, but forgot to remove it here. Removed in 0feebf99e40f59e23940d45b1f01d36131ba6fb7.